src/lib/hasher/solana.ts

```diff
@@ -0,0 +1,415 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import { createHash } from 'crypto'
+import { PublicKey } from '@solana/web3.js'
+import BN from 'bn.js'
+import { type Schema, serialize } from 'borsh'
+import type { CCIPMessage, CCIPVersion } from '../types.ts'
+import type { LeafHasher } from './common.ts'
+
+// Message types from https://github.com/smartcontractkit/chainlink-ccip/blob/main/chains/solana/contracts/programs/ccip-offramp/src/messages.rs
+/**
+ * RampMessageHeader
+ * - message_id: fixed 32 bytes array.
+ * - source_chain_selector, dest_chain_selector, sequence_number, nonce: u64 values.
+ */
+class RampMessageHeader {
+  message_id: Uint8Array
+  source_chain_selector: BN
+  dest_chain_selector: BN
+  sequence_number: BN
+  nonce: BN
+
+  constructor(fields: {
+    message_id: Uint8Array
+    source_chain_selector: BN
+    dest_chain_selector: BN
+    sequence_number: BN
+    nonce: BN
+  }) {
+    this.message_id = fields.message_id
+    this.source_chain_selector = fields.source_chain_selector
+    this.dest_chain_selector = fields.dest_chain_selector
+    this.sequence_number = fields.sequence_number
+    this.nonce = fields.nonce
+  }
+}
+
+/**
+ * Any2SVMRampExtraArgs
+ * - compute_units: u32
+ * - is_writable_bitmap: u64
+ */
+class Any2SVMRampExtraArgs {
+  compute_units: number
+  is_writable_bitmap: BN
+
+  constructor(fields: { compute_units: number; is_writable_bitmap: BN }) {
+    this.compute_units = fields.compute_units
+    this.is_writable_bitmap = fields.is_writable_bitmap
+  }
+}
+
+/**
+ * CrossChainAmount
+ * - le_bytes: fixed 32 bytes array.
+ */
+class CrossChainAmount {
+  le_bytes: Uint8Array
+
+  constructor(fields: { le_bytes: Uint8Array }) {
+    this.le_bytes = fields.le_bytes
+  }
+}
+
+/**
+ * Any2SVMTokenTransfer
+ * - source_pool_address: Vec<u8> as Uint8Array
+ * - dest_token_address: Pubkey (32 bytes) as Uint8Array
+ * - dest_gas_amount: u32
+ * - extra_data: Vec<u8> as Uint8Array
+ * - amount: CrossChainAmount
+ */
+class Any2SVMTokenTransfer {
+  source_pool_address: Uint8Array
+  dest_token_address: Uint8Array
+  dest_gas_amount: number
+  extra_data: Uint8Array
+  amount: CrossChainAmount
+
+  constructor(fields: {
+    source_pool_address: Uint8Array
+    dest_token_address: Uint8Array
+    dest_gas_amount: number
+    extra_data: Uint8Array
+    amount: CrossChainAmount
+  }) {
+    this.source_pool_address = fields.source_pool_address
+    this.dest_token_address = fields.dest_token_address
+    this.dest_gas_amount = fields.dest_gas_amount
+    this.extra_data = fields.extra_data
+    this.amount = fields.amount
+  }
+}
+
+/**
+ * Any2SVMRampMessage
+ * - header: RampMessageHeader
+ * - sender: Vec<u8> as Uint8Array
+ * - data: Vec<u8> as Uint8Array
+ * - token_receiver: Pubkey (32 bytes) as Uint8Array
+ * - token_amounts: array of Any2SVMTokenTransfer
+ * - extra_args: Any2SVMRampExtraArgs
+ */
+class Any2SVMRampMessage {
```

Comment:
This type should be mostly compatible with `CCIPMessage`; please, don't create a class only for it.
Also, the codebase is in functional style, and it's better not to use a class simply to hold some data, you can use a simple `interface` (or actually use the one we already have)